### DISCUSSION TITLE
Dedicated Matrix-Vector Multiplications Routines

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1535] [Q&A]
My goal is to implement a randomized SVD of a (block)-Hankel matrix and exploit the Hankel structure by using DFT-based matrix-Vector multiplications in order to enable ERA for high-dimensional data (s. https://epubs.siam.org/doi/10.1137/20M1327616).

I might need some help on how to achieve this in a "pymor-manner". I think it would be elegant to introduce a new operator, say `BlockHankelOperator`, and implement the Matrix-Vector routines inside the `apply` methods in the hopes that it nicely plugs into other methods. I already did something similar by subclassing `ndarray` in my existing non-pymor codebase.

Is this a feasible approach? If so, I will start working on a draft.

Similarly, many other routines e.g. for banded, triangular or Hessenberg matrixes could be implemented by using the scipy BLAS wrappers. Are there already ideas/opinions in these regards?

Is it a good idea to create a new randomized SVD routine that essentially combines the randrange finder with qr_svd?
A possible problem here could be that randrange accepts `Operator` inputs, whereas the svd methods expect `VectorArrays`

**Comment by pmli** [comment]
> My goal is to implement a randomized SVD of a (block)-Hankel matrix and exploit the Hankel structure by using DFT-based matrix-Vector multiplications in order to enable ERA for high-dimensional data (s. https://epubs.siam.org/doi/10.1137/20M1327616).
> 
> I might need some help on how to achieve this in a "pymor-manner". I think it would be elegant to introduce a new operator, say `BlockHankelOperator`, and implement the Matrix-Vector routines inside the `apply` methods in the hopes that it nicely plugs into other methods. I already did something similar by subclassing `ndarray` in my existing non-pymor codebase.
> 
> Is this a feasible approach? If so, I will start working on a draft.

It sounds like an interesting idea to me. I guess that the `apply` method would be similar to [`scipy.linalg.matmul_toeplitz`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.matmul_toeplitz.html)?

Instead of implementing a new `Operator` class, another option would be using a [`NumpyGenericOperator`](https://docs.pymor.org/2021-2-0/autoapi/pymor/operators/numpy/index.html#pymor.operators.numpy.NumpyGenericOperator), but then the source and range spaces would have to be `NumpyVectorSpaces`. So, actually, implementing `BlockHankelOperator` might only be worth the effort if there is a way to make it work with generic `VectorArrays`.

> Similarly, many other routines e.g. for banded, triangular or Hessenberg matrixes could be implemented by using the scipy BLAS wrappers. Are there already ideas/opinions in these regards?

Do you mean having generic `BandedOperator`, etc.? I guess that the fundamental issue would be that the `VectorArrays` would have to be NumPy-based to be able to use SciPy. This would be fine for reduced `Operators` (and could be interesting to have), but would not be "pyMOR-ic" in the generic case.

> Is it a good idea to create a new randomized SVD routine that essentially combines the randrange finder with qr_svd? A possible problem here could be that randrange accepts `Operator` inputs, whereas the svd methods expect `VectorArrays`

I'm not sure what you mean here. Do you want to have a randomized SVD routine for `Operators` instead of just a randomized range finder? I guess that it would be reasonable to have such a method.

**Reply by artpelling** [reply]
> It sounds like an interesting idea to me. I guess that the `apply` method would be similar to [`scipy.linalg.matmul_toeplitz`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.matmul_toeplitz.html)?
>
Exactly, I have made a small example in #1537.

> Instead of implementing a new `Operator` class, another option would be using a [`NumpyGenericOperator`](https://docs.pymor.org/2021-2-0/autoapi/pymor/operators/numpy/index.html#pymor.operators.numpy.NumpyGenericOperator), but then the source and range spaces would have to be `NumpyVectorSpaces`. So, actually, implementing `BlockHankelOperator` might only be worth the effort if there is a way to make it work with generic `VectorArrays`.

> Do you mean having generic `BandedOperator`, etc.? I guess that the fundamental issue would be that the `VectorArrays` would have to be NumPy-based to be able to use SciPy. This would be fine for reduced `Operators` (and could be interesting to have), but would not be "pyMOR-ic" in the generic case.
> 
For my application, just using numpy is absolutely fine. I have written some `ndarray` subclasses that use the SciPy BLAS routines in the past, but I think pyMOR `Operator`s are much nicer to work with. I cannot say much about the pyMOR-ic way. I could, however, see banded matrices come in handy e.g. when propagating linear difference equations with banded or tridiagonal `A` matrices.

> I'm not sure what you mean here. Do you want to have a randomized SVD routine for `Operators` instead of just a randomized range finder? I guess that it would be reasonable to have such a method.

Yes.

**Comment by sdrave** [comment]
> Is it a good idea to create a new randomized SVD routine that essentially combines the randrange finder with qr_svd?

Actually, I have an ongoing student project with the aim of implementing randomized SVD and related methods in pyMOR. I am hoping for a pr in the next weeks ..

**Reply by artpelling** [reply]
This is great news :-)
